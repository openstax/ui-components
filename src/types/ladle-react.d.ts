declare module '@ladle/react/serve';