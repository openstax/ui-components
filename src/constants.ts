export const maxNavWidth = 128;
export const navDesktopHeight = 4.0;
export const navMobileHeight = 4.0;
export const navLogoDesktopHeight = 2.4;
export const navLogoMobileHeight = 2.4;
