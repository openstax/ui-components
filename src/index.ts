export * from './components/Button';
export * from './components/Checkbox';
export * from './components/Error';
export * from './components/ErrorBoundary';
export * from './components/ErrorMessage';
export * from './components/ErrorModal';
export * from './components/Loader';
export * from './components/ManageCookies';
export * from './components/Modal';
export * from './components/NavBar';
export * from './components/NavBarLogo';
export * from './components/Overlay';
export * from './components/RiceLogo';
export * from './components/Text';
export * from './components/ToastContainer';
export * from './theme';
export * from './types';
