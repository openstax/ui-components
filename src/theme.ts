import { css, FlattenSimpleInterpolation } from "styled-components";
import { palette } from "./theme/palette";

export const colors = {
  palette: palette,
};

export const zIndex = [
  'navbar',
  'modals',
  'toasts'
].reduce((result, key, index) => {
  result[key] = (index + 1) * 10;
  return result;
}, {} as { [key: string]: number });

export const padding = {
  navbar: {
    desktop: 3.2,
    mobile: 1.6,
  },
};

const mobileBreak = 75; // 1200px
const desktopBreak = mobileBreak + .0625; // 1201px
const breakpoints = {
  desktop: (style: FlattenSimpleInterpolation) => css`
    @media screen and (min-width: ${desktopBreak}em) {
      ${style}
    }
  `,
};

const theme = {
  colors,
  zIndex,
  padding,
  breakpoints
};

export default theme; // eslint-disable-line import/no-default-export
