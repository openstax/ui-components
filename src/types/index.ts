export type ToastData = {
  id?: string;
  title: string;
  message: React.ReactNode;
  variant?: ToastVariant;
  dismissAfterMs?: number;
  onDismiss?: (id: ToastData['id']) => void;
}

type ToastVariant = 'neutral' | 'success' | 'failure';

export type SentryError = {
  error: Error;
  componentStack?: string;
  eventId?: string;
  type?: string;
};
