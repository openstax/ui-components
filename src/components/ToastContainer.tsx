import { zIndexes } from '../../src/theme';
import styled from 'styled-components';
import { Toast } from './Toast';
import { ToastData } from '../../src/types';

const StyledToastContainer = styled.div`
  position: fixed;
  right: 2rem;
  z-index: ${zIndexes.toasts};
  display: grid;
  justify-items: center;
  justify-content: center;
  gap: 1vh;
`;

export const ToastContainer = ({ toasts, onRemoveToast }: {
  toasts: ToastData[], onRemoveToast?: ToastData['onDismiss']
}) => {
  return <StyledToastContainer>
    {toasts.map((toast, index) => <Toast
      key={`toast-${index}`}
      onDismiss={(id) => onRemoveToast ? onRemoveToast(id) : null}
      {...toast}
      >{toast.message}</Toast>)}
  </StyledToastContainer>
}
