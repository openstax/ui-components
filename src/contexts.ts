import React from 'react';
import { SentryError } from './types';

export const ErrorContext = React.createContext<{
  error: SentryError | null;
  setError: React.Dispatch<React.SetStateAction<SentryError | null>>;
}>({
  error: null,
  setError: () => {
    throw new Error('Expected setError to be initialized');
  }
});
