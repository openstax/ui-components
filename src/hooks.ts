import React from 'react';
import { ErrorContext } from './contexts';
import { getTypeFromError } from './utils';

export const useSetAppError = () => {
  const { setError } = React.useContext(ErrorContext);

  return React.useCallback((error: Error | null) => {
    setError(
      error ? {
        error: error,
        type: getTypeFromError(error),
      } : null
    );
  }, [setError])
}
