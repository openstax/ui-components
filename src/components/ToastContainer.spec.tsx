import { ToastContainer } from './ToastContainer';
import renderer from 'react-test-renderer';
import { ToastData } from '../../src/types';

jest.useFakeTimers();

const toasts: ToastData[] = [
  { id: '1', title: 'Neutral', message: 'message' },
  { id: '2', title: 'Error', message: 'message', variant: 'failure', dismissAfterMs: 4000 },
  { id: '3', title: 'Success', message: 'message', variant: 'success', dismissAfterMs: 2000 },
];

describe('ToastContainer', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(<ToastContainer toasts={toasts} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('runs callback', () => {
    const callback = jest.fn();
    renderer.create(<ToastContainer toasts={toasts.splice(0)} onDismissToast={callback} />).toJSON();
    renderer.act(() => {
      jest.advanceTimersByTime(10000);
      expect(callback).toHaveBeenCalledWith('3')
      expect(callback).toHaveBeenCalledWith('2')
    });
  });
});
