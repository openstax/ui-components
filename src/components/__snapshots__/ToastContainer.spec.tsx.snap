// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastContainer matches snapshot 1`] = `
<body>
  <div
    class="sc-gsnTZi bQKHKA"
    data-portal-slot="toast"
  >
    <div
      class="sc-bczRLJ jAZNhq"
    >
      <div
        class="neutral"
      >
        <div
          class="title"
        >
          Neutral
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
    <div
      class="sc-bczRLJ cWrNHg"
    >
      <div
        class="failure"
      >
        <div
          class="title"
        >
          Error
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
    <div
      class="sc-bczRLJ loazPe"
    >
      <div
        class="success"
      >
        <div
          class="title"
        >
          Success
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
  </div>
  <main
    id="root"
  />
</body>
`;

exports[`ToastContainer uses inline prop 1`] = `
<body>
  <div
    class="sc-gsnTZi gRYEVl"
    data-portal-slot="toast"
  >
    <div
      class="sc-bczRLJ jSNerL"
    >
      <div
        class="neutral"
      >
        <div
          class="title"
        >
          Neutral
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
    <div
      class="sc-bczRLJ fgCwHl"
    >
      <div
        class="failure"
      >
        <div
          class="title"
        >
          Error
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
    <div
      class="sc-bczRLJ dGEkoL"
    >
      <div
        class="success"
      >
        <div
          class="title"
        >
          Success
        </div>
        <div
          class="body"
        >
          message
        </div>
      </div>
    </div>
  </div>
  <main
    id="root"
  />
</body>
`;
